kernel/mod_mgmt/src/lib.rs

```diff
@@ -161,36 +171,75 @@ fn parse_bootloader_modules_into_files(
+            return Ok(());
         };
 
         let dir_name = format!("{}{}", prefix, crate_type.default_namespace_name());
         // debug!("Module: {:?}, size {}, mp: {:?}", m.name(), m.size_in_bytes(), mp);
 
         let create_file = |dir: &DirRef| {
-            MemFile::from_mapped_pages(mp, file_name.to_string(), m.size_in_bytes(), dir)
+            MemFile::from_mapped_pages(pages, file_name.to_string(), size, dir)
         };
-
         // Get the existing (or create a new) namespace directory corresponding to the given directory name.
         let _new_file = match prefix_map.entry(dir_name.clone()) {
             btree_map::Entry::Vacant(vacant) => create_file( vacant.insert(create_dir(&dir_name)?) )?,
             btree_map::Entry::Occupied(occ)  => create_file( occ.get() )?,
         };
+        Ok(())
+    };
+
+    for m in bootloader_modules {
+        let frames = allocate_frames_by_bytes_at(m.start_address(), m.size_in_bytes())
+            .map_err(|_e| "Failed to allocate frames for bootloader module")?;
+        let pages = allocate_pages_by_bytes(m.size_in_bytes())
+            .ok_or("Couldn't allocate virtual pages for bootloader module area")?;
+        let mp = kernel_mmi.page_table.map_allocated_pages_to(
+            pages,
+            frames,
+            // we never need to write to bootloader-provided modules
+            EntryFlags::PRESENT | EntryFlags::NO_EXECUTE,
+        )?;
+
+        let name = m.name().clone();
+        let size = m.size_in_bytes();
+
+        if name.as_str() == "modules.cpio.lz4" {
+            // modules are zipped
+
+            #[cfg(feature = "mod_unpack")]
+            {
+                let bytes = mp.as_slice(0, size)?;
+                let tar = decompress_size_prepended(bytes).ok()
+                    .ok_or("lz4 decompression failed")?;
+                for entry in iter_files(&tar) {
+                    let name = entry.name().to_string();
+                    let bytes = entry.file();
+                    let size = bytes.len();
+                    debug!("extracting {} ({}B)", &name, size);
+                    let mut mp = {
+                        let err_msg = "couldn't allocate pages";
+                        let flags = EntryFlags::WRITABLE | EntryFlags::NO_EXECUTE | EntryFlags::PRESENT;
+                        let allocated_pages = allocate_pages_by_bytes(size).ok_or(err_msg)?;
+                        kernel_mmi.page_table.map_allocated_pages(allocated_pages, flags)?
+                    };
+                    {
+                        let slice = mp.as_slice_mut(0, size)?;
+                        slice.copy_from_slice(bytes);
+                    }
+                    process_module(name, size, mp)?;
+                }
+                continue;
+            }
+            #[cfg(not(feature = "mod_unpack"))]
+            debug!("modmgmt: WARNING! `modules.cpio.lz4` encountered but the `mod_unpack` feature is disabled!");
```

Comment:
this should be a hard error, right? you can both log an `error!()` message and return an `Err`.

Reply:
yep, done